### PR TITLE
Implement all

### PR DESCRIPTION
**Pull request by Resci**

### CODE REVIEW

**Review by AndriiBardin — APPROVED** [review]

File: src/main/java/com/mate/exception/AuthenticationException.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.mate.exception;
+
+public class AuthenticationException extends RuntimeException {
```
Comment (KseniiaMakarova) [inline src/main/java/com/mate/exception/AuthenticationException.java]: should it be Runtime?

File: src/main/java/com/mate/util/PasswordUtils.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Random;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+ 
+public class PasswordUtils {
```
Comment (KseniiaMakarova) [inline src/main/java/com/mate/util/PasswordUtils.java]:
```suggestion
public class PasswordUtil {
```

File: src/main/java/com/mate/util/PasswordUtils.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/mate/util/PasswordUtils.java]: redundant empty line

File: src/main/java/com/mate/util/PasswordUtils.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            + "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+    private static final int ITERATIONS = 10000;
+    private static final int KEY_LENGTH = 256;
+    
+    public static String getSalt(int length) {
+        StringBuilder returnValue = new StringBuilder(length);
+        for (int i = 0; i < length; i++) {
+            returnValue.append(ALPHABET.charAt(RANDOM.nextInt(ALPHABET.length())));
+        }
+        return new String(returnValue);
```
Comment (KseniiaMakarova) [inline src/main/java/com/mate/util/PasswordUtils.java]: why can't we use SecureRandom?

File: src/main/java/com/mate/util/PasswordUtils.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public static byte[] hash(char[] password, byte[] salt) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/mate/util/PasswordUtils.java]: should this method be private?

File: src/main/java/com/mate/util/PasswordUtils.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public static String generateSecurePassword(String password, String salt) {
+        String returnValue = null;
+        byte[] securePassword = hash(password.toCharArray(), salt.getBytes());
+        returnValue = Base64.getEncoder().encodeToString(securePassword);
+        return returnValue;
+    }
+    
+    public static boolean verifyUserPassword(String providedPassword,
```
Comment (KseniiaMakarova) [inline src/main/java/com/mate/util/PasswordUtils.java]: do we need this method or can we do this logic in `login()`?

File: src/main/java/com/mate/util/PasswordUtils.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            String securedPassword, String salt) {
+        boolean returnValue = false;
+        String newSecurePassword = generateSecurePassword(providedPassword, salt);
+        returnValue = newSecurePassword.equalsIgnoreCase(securedPassword);
```
Comment (KseniiaMakarova) [inline src/main/java/com/mate/util/PasswordUtils.java]: I think `equals()` fits better here - passwords should be exactly the same

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/mate/util/PasswordUtil.java [hunk 8]
```diff
@@ -9,23 +9,18 @@
...
-            returnValue.append(ALPHABET.charAt(RANDOM.nextInt(ALPHABET.length())));
-        }
-        return new String(returnValue);
+        byte[] salt = new byte[length];
+        RANDOM.nextBytes(salt);
+        return Base64.getEncoder().encodeToString(salt);
     }
 
-    public static byte[] hash(char[] password, byte[] salt) {
+    private static byte[] hash(char[] password, byte[] salt) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/mate/util/PasswordUtil.java:28]: should private method be located below its usages?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**